prov/mxm/src/mlxm.h

```diff
@@ -0,0 +1,306 @@
+void mlxm_mq_storage_init() {
+        mlxm_globals.mq_storage.hash = NULL;
+}
+
+static inline
+void mlxm_mq_storage_fini(mlxm_fid_ep_t *fid_ep) {
+        struct mlxm_mq_entry *mq_e, *tmp;
+        HASH_ITER(hh, mlxm_globals.mq_storage.hash, mq_e, tmp) {
+                mxm_mq_destroy(mq_e->mq);
+                HASH_DEL(mlxm_globals.mq_storage.hash,
+                         mq_e);
+                free(mq_e);
+        }
+}
+
+struct mlxm_fid_domain {
+	struct fid_domain	domain;
+        size_t		        mxm_addrlen;
+};
+
+struct mlxm_cq_entry_queue {
+	size_t			size;
+	size_t			n;
+	struct mlxm_cq_entry	*head;
+	struct mlxm_cq_entry	*tail;
+};
+
+struct mlxm_fid_cq {
+        struct fid_cq cq;
+        mxm_h mxm_context;
+        struct {
+                void *head;
+                void *tail;
+        } ok_q;
+        struct {
+                void *head;
+                void *tail;
+        } err_q;
+};
+
+#define MLXM_CQ_ENQUEUE(_queue, __ctx)                                  \
+        do{                                                             \
+                if (_queue.head == NULL) {                              \
+                        _queue.head = __ctx->internal[0];               \
+                        _queue.tail = __ctx->internal[0];               \
+                } else {                                                \
+                        ((struct fi_context *)(_queue.tail))->internal[0] = \
+                                __ctx;                                  \
+                        _queue.tail = __ctx;                            \
+                }                                                       \
+        }while(0)
+
+#define MLXM_CQ_DEQUEUE(_queue, __ctx)                          \
+        do{                                                     \
+                __ctx = (struct fi_context*)(_queue.head);      \
+                assert(__ctx);                                  \
+                if (__ctx->internal[0] == __ctx) {              \
+                        _queue.head = NULL;                     \
+                } else {                                        \
+                        _queue.head = __ctx->internal[0];       \
+                }                                               \
+        }while(0)
+
+struct mlxm_fid_av {
+	struct fid_av		av;
+        mlxm_fid_domain_t	*domain;
+        mlxm_fid_ep_t   	*ep;
+	int			type;
+	size_t count;
+        size_t			addrlen;
+};
+
+struct mlxm_fid_ep {
+	struct fid_ep		ep;
+        mlxm_fid_domain_t	*domain;
+        mlxm_fid_cq_t   	*cq;
+        mlxm_fid_av_t   	*av;
+	uint64_t		flags;
+        struct mlxm_mq_storage*	mxm_mqs;
+};
+
+struct mlxm_fid_fabric {
+	struct fid_fabric	fabric;
+};
+
+typedef struct mlxm_fid_mr {
+	struct fid_mr		mr;
+        struct mlxm_fid_domain	*domain;
+        mxm_mem_key_t mxm_key;
+        size_t iov_count;
+        struct iovec		iov[0];	/* must be the last field */
+}mlxm_fid_mr_t;
```

Comment:
Do u mean I should remove all typedefs? I thought it's ok to use them to save some typing (i've seen psm provider declaring struct with a typedef)